File: problems/865_smallest_subtree_with_all_the_deepest_nodes/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        return dfs(root).first;
    }
    pair<TreeNode*, int> dfs(TreeNode* node) {
        if (!node) return {nullptr, 0};
        auto [ln, ld] = dfs(node->left);
        auto [rn, rd] = dfs(node->right);
        if (ld > rd) return {ln, ld + 1};
        if (rd > ld) return {rn, rd + 1};
        return {node, ld + 1};
    }
};

File: problems/865_smallest_subtree_with_all_the_deepest_nodes/Solution.java
class Solution {
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        return dfs(root)[0];
    }
    TreeNode[] dfs(TreeNode node) {
        if (node == null) return new TreeNode[]{null, null};
        TreeNode[] left = dfs(node.left), right = dfs(node.right);
        int ld = depth(left), rd = depth(right);
        if (ld > rd) return new TreeNode[]{left[0], wrap(ld + 1)};
        if (rd > ld) return new TreeNode[]{right[0], wrap(rd + 1)};
        return new TreeNode[]{node, wrap(ld + 1)};
    }
    int depth(TreeNode[] arr) { return arr[1] == null ? 0 : arr[1].val; }
    TreeNode wrap(int d) { TreeNode n = new TreeNode(d); return n; }
}
